### PR TITLE
Gear 505 add dicom support

### PR DESCRIPTION
**Pull request by joshicola**
* Embedded dicom-to-nifti (dcmpi) converts DICOM series to NIfTI file
  * Will FAIL if DICOM Tags are not present:
    * "SeriesInstanceUID", "InstanceNumber", "ImageOrientationPatient", "ImagePositionPatient" 
* Axial, Coronal, and Sagittal orientations to DICOM are accommodated.
* A NIfTI representation of the DICOM (`converted_dicom.nii.gz`) is provided as output to compare with ROI output.
  * Some modifications may be necessary to accommodate for the errors above.

### CODE REVIEW

File: utils.py [hunk 1]
```diff
@@ -5,17 +5,247 @@
...
+):
+    """
+    Converts ohifViewer ROI from a dicom to that of an individual nifti format.
+
+    Args:
+        fw_client (flywheel.Client): The active flywheel client
+        project_id (str): The id of the project to search for io-proxy elements
+        file_obj (flywheel.File): File object data
+        imagePath (str): The DICOM representation of the ohifViewer ROI imagePath
+        ohifViewer_info (dict): The full session-level ohifViewer ROI dictionary.
```
Comment (AndysWorth) [inline utils.py:104]: Maybe explain why this needs to be passed in here.  I expect that it has higher level ROI info and then `convert_ROI_to_nifti_form()` fills in the actual outlines?  This is probably my only chance to ask big-picture questions.

File: utils.py [hunk 2]
```diff
@@ -5,17 +5,247 @@
...
+    study_id, series_id = imagePath.split("$$$")[:2]
+
+    instances = io_proxy_wado(api_key, api_key_prefix, project_id, study_id, series_id)
+
+    for roi_type in ["FreehandRoi", "RectangleRoi", "EllipticalRoi"]:
+        if ohifViewer_info["measurements"].get(roi_type):
+            for roi in ohifViewer_info["measurements"].get(roi_type):
+                if imagePath in roi["imagePath"]:
+                    instance_id = roi["imagePath"].split("$$$")[2]
+                    slice_instance = [
```
Comment (AndysWorth) [inline utils.py:127]: So this gets the good stuff?  Maybe include a description of what is being gotten?

File: utils.py [hunk 3]
```diff
@@ -5,17 +5,247 @@
...
+                    if roi_type not in out_ohifViewer_info["measurements"].keys():
+                        out_ohifViewer_info["measurements"][roi_type] = []
+
+                    # (0020, 0037) Image Orientation Patient
+                    # Needed to determine axis perpendicular to slice
+                    ImageOrientationPatient = list(
+                        map(round, slice_instance["00200037"]["Value"])
+                    )
+                    # from https://stackoverflow.com/questions/34782409/understanding-dicom-image-attributes-to-get-axial-coronal-sagittal-cuts
+                    # https://groups.google.com/g/comp.protocols.dicom/c/GW04ODKR7Tc?pli=1
```
Comment (AndysWorth) [inline utils.py:142]: Yay, sources are always good!

**Review by AndysWorth — APPROVED** [review]
This looks great.  I'm mostly asking for clarification so when I need to come back to this, it won't take forever to understand the trouble you had to go through to get here.  Merge on!

File: run.py [hunk 4]
```diff
@@ -29,18 +29,34 @@ def main(context):
...
+        elif file_obj["type"] == "dicom":
+            # convert dicom-centric data to nifti-centric data
+            nii, file_obj, perp_char = convert_dicom_to_nifti(
+                context, input_name="Input_File"
+            )
+            # If the DICOM is Axial or Coronal (perp_char is "z" or "y") modify the
+            # below adjustment matrix to indicate the left/right origin of the x-axis.
+            if perp_char in ["z", "y"]:
+                adjustment_matrix[:, 0] = -1 * adjustment_matrix[:, 0]
+
```
[the remark this answers is not in the document]
Comment (joshicola) [reply run.py:46]: This particular "adjustment" only affects the views with a L/R axis (axial, coronal).

File: utils.py (hunk 1 above)
Comment (joshicola) [reply utils.py:104]: The current ohifViewer annotations on dicoms end up in the session info.

Comment (joshicola) [reply utils.py:104]: Questions after next week will be $$$$.

File: utils.py [hunk 5]
```diff
@@ -5,17 +5,247 @@
...
+    Args:
+        fw_client (flywheel.Client): The active flywheel client
+        project_id (str): The id of the project to search for io-proxy elements
+        file_obj (flywheel.File): File object data
+        imagePath (str): The DICOM representation of the ohifViewer ROI imagePath
+        ohifViewer_info (dict): The full session-level ohifViewer ROI dictionary.
+
+    Returns:
+        dict: File object (file_obj) dictionary
+    """
```
[the remark this answers is not in the document]
Comment (joshicola) [reply utils.py:108]: Good Catch!

File: utils.py (hunk 2 above)
Comment (joshicola) [reply utils.py:127]: Good idea!!

File: utils.py [hunk 6]
```diff
@@ -5,17 +5,247 @@
...
+    # Get configuration, acquisition, and file info
+    file_input = context.get_input(input_name)
+
+    # Need updated file information.
+    file_obj = file_input["object"]
+    # ASSUMPTION: DICOMs will be stored at acquisition level
+    acquisition = fw_client.get(file_input["hierarchy"]["id"])
+    project_id = acquisition.parents["project"]
+    # Prioritize dicom file-level ROI annotations
+    #   -- if they should occur this way soon....
```
[the remark this answers is not in the document]
Comment (joshicola) [reply utils.py:174]: This is preparation for a future feature: DICOM annotations associated with file-info.

[the remark this answers is not in the document]
Comment (joshicola) [reply utils.py:174]: Noted.

File: utils.py [hunk 7]
```diff
@@ -5,17 +5,247 @@
...
+
+    # open one dicom file to extract studyInstanceUid and seriesInstanceUid
+    dicom = None
+    for root, _, files in os.walk(str(dicom_dir), topdown=False):
+        for fl in files:
+            try:
+                dicom_path = Path(root) / fl
+                dicom = pydicom.read_file(dicom_path, force=True)
+                break
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (joshicola) [reply utils.py:209]: Noted.
Trying until we get a dicom file... that we can open and get the tags within.

File: utils.py [hunk 8]
```diff
@@ -5,17 +5,247 @@
...
+        Exception.__init__(self)
+        self.message = message
+
+
+def io_proxy_wado(
+    api_key, api_key_prefix, project_id, study=None, series=None, instance=None
+):
+    """
+    Request wrapper for io-proxy api (https://{instance}/io-proxy/docs#/)
+
```
[the remark this answers is not in the document]
Comment (joshicola) [reply utils.py:58]: There is a "wado" database that keeps track of all the dicom information.

File: run.py [hunk 9]
```diff
@@ -29,18 +29,34 @@ def main(context):
...
         # Create an inverse of the matrix that is the closest projection onto the
         # basis unit vectors of the coordinate system of the original affine.
-        inv_reduced_aff = np.linalg.inv(
-            np.round(
-                np.matmul(
-                    nii.affine[:3, :3],
-                    np.diag(1.0 / np.linalg.norm(nii.affine[:3, :3], axis=0)),
+        # This is used to determine which axes to flip
```
[the remark this answers is not in the document]
Comment (joshicola) [reply run.py:49]: Added some notes.

**Comment by joshicola** [comment]
Thanks for the review, @AndysWorth !  I added some notes to clarify your points.

**Pull request merged**